warehouse/query-core/src/main/java/datawave/query/language/processor/lucene/CustomAnalyzerQueryNodeProcessor.java

```diff
@@ -408,4 +438,68 @@ private String toStringEscaped(UnescapedCharSequence unescaped) {
         }
         return result.toString();
     }
+
+    /**
+     * Maintains one or more buffers for tokenized queries. During standard operation, works like a StringBuilder. If the tokenizer encounters a variant (e.g.,
+     * zero position offset, same start and end as the previous token) appendVariant will start building a second buffer containing that variant.
+     */
+    public static class VariantBuilder {
+        List<List<String>> variants = new ArrayList<>();
+
+        public VariantBuilder append(String input, boolean appendVariant) {
+            return appendVariant ? appendVariant(input) : append(input);
+        }
+
+        public VariantBuilder append(String input) {
+            if (variants.isEmpty()) {
+                variants.add(new ArrayList<>());
+            }
+
+            for (List<String> b : variants) {
+                b.add(input);
+            }
+
+            return this;
+        }
+
+        public VariantBuilder appendVariant(String input) {
+            if (variants.isEmpty()) {
+                append(input);
+            } else {
+
+                List<List<String>> newVariants = new ArrayList<>();
+
+                for (List<String> b : variants) {
+                    // create a new variant of all the existing strings, replacing the
+                    List<String> newVariant = new ArrayList<>(b);
```

Comment:
Since there is no random access, might it make sense to use a a LinkedList rather than an array list to prevent reallocating the list for each new variant? It seems like this could get expensive.